### PR TITLE
update to v1.41.0

### PR DESCRIPTION
**Pull request by bbara**
Hi!

I fetched the git diff from v1.18.0 to v1.41.0 of the Makefile.am and added the respective changes (linux/apple) to the meson.build.

BR
bb

### CODE REVIEW

**Comment by jpakkane** [comment]
MR retargeted. Please squash the commits down to one.

**Comment by bbara** [comment]
@jpakkane done

File: upstream.wrap
Comment (jpakkane) [inline upstream.wrap:7]: Nowadays all wraps need a `provide` section:

https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section

Comment (bbara) [reply upstream.wrap:7]: done

**Comment by eli-schwartz** [comment]
The tests fail unless you set workdir to `meson.current_source_dir()` in which case they're reported as skipped.

They seem to be using TAP format for reporting, but for some reason adding `protocol: 'tap'` does not cause meson to report the subtests...

**Comment by bbara** [comment]
true. I take a look into that

**Comment by bbara** [comment]
@eli-schwartz, finally got little time to debug it:
the libuv test runner writes the TAP output to stderr, while meson is expecting it on stdout.
I think it would be a good idea for mtest to additionally consider stderr (or only if stdout is empty) - what do you think about it?

**Comment by eli-schwartz** [comment]
That's definitely not something I'd expect, lol.

Reading from stderr is non-negotiable, as the TAP specification explicitly forbids this.

**Comment by bbara** [comment]
Thx for the feedback - just found it in the spec. I will file a respective PR in the libuv project and will update this PR accordingly with the new version as soon as it is available. For now, I set it to WIP until this is done.

File: meson.build
Comment (eli-schwartz) [inline meson.build]: This is a problem with the previous released version of the wrap too, but... how does this work, exactly? This is a stub for Unix, there is also a `test/runner-win.c`, and there's no way you can unconditionally include this... the tests run fine here only because I'm a Linux user, I think. ;)

This should include one or the other runner depending on the OS.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: meson.build
Comment (bbara) [reply meson.build]: The initial wrap completely ignores win32 (and other platforms) at all and is just a small working subset of the actual CMake/autoconf. I can do some improvements s.t. the meson.build comes closer to the CMakeLists.txt (which I think didn't exist when the initial wrap was created).

Comment (eli-schwartz) [reply meson.build]: I didn't check the history, but there are *currently* `if WINNT` blocks in Makefile.am, too.

Either way it would be quite good to have...

Comment (bbara) [reply meson.build]: I'm just wondering if it is worth the work, because the project's maintainers seem to be not interested in supporting meson as additional build system (just found https://github.com/libuv/libuv/pull/3075) and with the CMake file in place, it now should also be easily integrateable into meson projects.

Comment (eli-schwartz) [reply meson.build]: https://mesonbuild.com/External-Project-module.html supports the `./configure; make` build system.

CMake projects are not distributable via the wrapdb, and require one to have both meson and cmake installed, so yes, ideally there is a native meson.build wrap.

Comment (bbara) [reply meson.build]: nah, i'll do it - why not. maybe they reconsider their decision in the future, then they'll have a nice starting point. follow up is coming soon :)

File: meson_options.txt
Comment (eli-schwartz) [inline meson_options.txt]: Project options don't need to include the word "libuv" just because cmake -Doption sets a globally scoped variable in the CMakeLists.txt; also capitalizing the options is less common in meson.

Why does this mention "we are the root project"?

File: meson.build
Comment (eli-schwartz) [inline meson.build]: Do these need such large timeouts? Probably not.

Seems disappointing if they don't run in parallel. :(

File: meson.build
Comment (eli-schwartz) [inline meson.build]: As a matter of curiosity -- why explicitly build both static and shared libs, and test both of them? Shouldn't people be using default_library to select one or the other?

**Review by eli-schwartz — COMMENTED** [review]
Thanks for the update. Some preliminary comments below.

File: meson_options.txt
Comment (bbara) [reply meson_options.txt]: In the original cmake, these are cmake_dependent_option with the requirement 
`CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR`

As you pointed out earlier, I think this is required because of some hardcoded path dependencies in the tests, which we solve later by setting the workdir of the test, so these docs are basically copy/pasted and the dependency is not required anymore.

File: meson.build
Comment (bbara) [reply meson.build]: Sure, I can lower them.

yeah, I tried to run them in parallel but some tests are binding to ports etc and therefore a couple of EADDRINUSE etc. occur.

File: meson.build
Comment (bbara) [reply meson.build]: I basically followed the original cmake, because from my point of view, the two variants aren't using the exact same code, because the shared variant sets two defines: USING_UV_SHARED and BUILDING_UV_SHARED.

I just checked and it seems to be because of win32-specific stuff (`__declspec(dllexport), __declspec(dllimport)`).

Comment (eli-schwartz) [reply meson.build]: That's unfortunately a well-known problem with Microsoft. It might be only needed by msvc, not by the MinGW gcc.

There is an open issue for meson to support setting different c_args on the static/shared parts of a library()/both_libraries() but no merged code.

Another solution is to get_option('default_library'), choose which defines to use... and if the value is "both", just raise error('default_library=both is not supported on Windows').

Comment (bbara) [reply meson.build]: did that now, lgtm tbh

**Comment by jpakkane** [comment]
Does this need anything else besides squashing to a single commit?

**Comment by eli-schwartz** [comment]
It's not clear to me why the timeout is 300 instead of simply declaring nothing and using the meson default of 30.

Other than that, maybe we want to wait for the libuv upstream PR to be merged and released, that makes the testsuite actually be parseable by the TAP runner?

File: meson.build
Comment (eli-schwartz) [reply meson.build]: Why is 300 better than 600? The default is 30, do these need longer 10x or 20x longer? Your response doesn't make it clear if you've timed them and come up with an average reasonable amount of time for them to take...

**Comment by bbara** [comment]
it takes ~35s to run them on my machine, therefore 30s should be too low.
Not sure what is an acceptable threshold, 300 was the old value - i can set it to 60?
The libuv tests also have a timeout implemented (5s per test x 394 tests).

**Comment by jpakkane** [comment]
Let's do 100 seconds and squash and this is ready to go.

**Comment by bbara** [comment]
As @eli-schwartz said, the tests are not recognized by meson until https://github.com/libuv/libuv/pull/3153 is merged and released.

A side question: What do you think about setting setup a pipeline on some meson infrastructure (Github Actions, Jenkins, ...) which gets triggered as soon as the respective wrapdb project pushes a new tag?
Because on mature projects it might be quite rare that the project structure is changed, so if the respective build files (Makefiles, CMake, autotools, ..; should be parameters per project) are not changed since the last release, the existing meson.build could be reused and an automatic PR could be filed (maybe with the test results attached or something like that).

**Comment by jpakkane** [comment]
> As @eli-schwartz said, the tests are not recognized by meson until libuv/libuv#3153 is merged and released.

Is there an estimate on when that will be? If it takes a long time (weeks?) then we could release the current one without `protocol: 'tap'` as it is already usable (perfect being the enemy of good and all that).

**Comment by eli-schwartz** [comment]
Uhhh. It's been 10 days now so this is not looking too good in terms of timeframe if no one can get an estimate.

Currently the tests return SKIPPED because no TAP output, removing the protocol kwarg makes it return OK. Please do so (and leave a comment pointing to the upstream PR).

I agree that that is enough for now and given we have the promise of a proper fix in the future, we can just merge it like that.

**Comment by jpakkane** [comment]
It appears that the build files use tabs for alignment. This is not permitted, only spaces. You might want to use the [Mesonwrap tools](https://github.com/mesonbuild/mesonwrap/) to validate things (that is what reviewers also use). See [the docs](https://mesonbuild.com/Wrap-review-guidelines.html#setting-up-the-review-tooling) but basically:

- install `mesonwrap`
- run `mesonwrap review --pull-request 5 libuv`
- fix all issues reported

When all checks pass, this can be merged.

**Comment by bbara** [comment]
cool tool! tests show OK now and mesonwrap shows only checkmarks and no errors (but returns exit code 1).

File: meson.build
Comment (jpakkane) [inline meson.build]: You should not add `-Wall` or `/W4` by hand. They are added by Meson automatically based on the warning level option.

**Comment by eli-schwartz** [comment]
It just got merged upstream, only another 11 days later. :D

Now for a release...

**Pull request merged**
